cellpack/autopack/loaders/recipe_loader.py

```diff
@@ -165,8 +166,102 @@ def _migrate_version(self, old_recipe):
+        # TODO: true for all cases? better approaches?
+        return not comp_or_obj.get("name") and "object" in comp_or_obj
+
+    @staticmethod
+    def _collect_and_sort_data(comp_data):
+        """
+        Collect all object and gradient info from the downloaded firebase composition data
+        Return autopack object data dict and gradient data dict with name as key
+        Return restructured composition dict with "composition" as key
+        """
+        objects = {}
+        gradients = {}
+        composition = {}
+        for comp_name, comp_value in comp_data.items():
+            composition[comp_name] = {}
+            if "count" in comp_value and comp_value["count"] is not None:
+                composition[comp_name]["count"] = comp_value["count"]
+            if "object" in comp_value and comp_value["object"] is not None:
+                composition[comp_name]["object"] = comp_value["object"]["name"]
+                object_copy = copy.deepcopy(comp_value["object"])
+                objects[object_copy["name"]] = object_copy
+                if "gradient" in object_copy and isinstance(
+                    object_copy["gradient"], dict
+                ):
+                    objects, gradients = RecipeLoader._get_grad_and_obj(
+                        object_copy, objects, gradients
+                    )
+            if "regions" in comp_value and comp_value["regions"] is not None:
+                for region_name in comp_value["regions"]:
+                    composition[comp_name].setdefault("regions", {})[region_name] = []
+                    for region_item in comp_value["regions"][region_name]:
+                        if RecipeLoader._is_obj(region_item):
+                            composition[comp_name]["regions"][region_name].append(
+                                {
+                                    "object": region_item["object"].get("name"),
+                                    "count": region_item.get("count"),
+                                }
+                            )
+                            object_copy = copy.deepcopy(region_item["object"])
+                            objects[object_copy["name"]] = object_copy
+                            if "gradient" in object_copy and isinstance(
+                                object_copy["gradient"], dict
+                            ):
+                                objects, gradients = RecipeLoader._get_grad_and_obj(
+                                    object_copy, objects, gradients
+                                )
+                        else:
+                            composition[comp_name]["regions"][region_name].append(
+                                region_item["name"]
+                            )
+        return objects, gradients, composition
+
+    @staticmethod
+    def _compile_recipe_from_firebase(db_recipe_data, obj_dict, grad_dict, comp_dict):
```

Comment:
Shouldn't have anything referencing firebase in recipe loader. this should be in the database handler.